### PR TITLE
MRI Super-Resolution Network - Mike Smith (46408789)

### PR DESCRIPTION
**Pull request by mikesmith012**
This project involves creating a Super-Resolution Deep Convolutional Generative Adversarial Network (SR-DCGAN) to enhance low-resolution brain MRI images from the ADNI dataset. This network aims to accurately perform 4x upsampling to construct high-resolution images from sparse sampled, low-resolution inputs. This process is useful for restoring detailed structures that are otherwise lost through undersampling.

Included folders and files:
- `dataset.py`: Loads in images from the ADNI MRI dataset.
- `modules.py`: Contains the model structures for the Generator and the Discriminator models.
- `train.py`: Contrains the training loop to train the generator and discriminator models using the training dataset.
- `predict.py`: Makes predictions based on a trained model in the testing dataset.
- `README.md`: An in-depth description of the aims, models, training and results.
- `figures/`: Figures used in the README.
- `outputs/`: Text outputs from the training process and a python script for plotting / visualising outputs.

Overall, the generator model was able to produce "reasonably clear images", as shown in the README, reconstructed from 4x downsampled images.

Mike Smith
46408789

### CODE REVIEW

**Comment by uqmarlonbran** [comment]
**This is an initial inspection - no action is required yet**

**Difficulty:** Normal

**General Comments:** Good implementation of a conditional GAN super resolution network. Reconstructed images seem to have more detail than the low-rez versions. Code was easy to follow, though, evaluation of some metrics would have been nice.

**[Recognition Problem]**
"modules.py":

- Custom super-rez network implemented using ConvTranspose2d and GAN approach.
- Uses 2x upsampling blocks rather than 1 - probably better for perf.
- All looks ok implementation-wise.

"dataset.py":

- Nice use of ImageFolder to deal with AD and NC folders.
- Seems to use 3 channels rather than converting to grayscale.
- Normalizes images, but correctly un-normalizes in "predict.py".

"train.py":

- Seems to be a good implementation of GAN training.
- Model at the end of training procedure is used.
- Correctly tracks generator and discriminator losses.
- However, this is a problem that does have ground truth images, therefore some check to the MSE between generated images and ground truth images should probably be made.

"predict.py":

- Predict correctly loads network weights and visualises some predicitons.
- Again, no metrics on the test dataset are run.

"README.MD":

- Good description and figures of proposed model architecture for both generator and discriminator.
- Provides training losses, as well as learning rate schedules for reproducibility.
- Provides dependencies.
- Images seem of decent quality, particularly for a purely GAN training scheme.

**[Commit Log]**
Commit log looks ok.

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation 
Good spacing and comments
Header blocks missing -1
## Recognition Problem 
Solves problem 
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage
Module present
Commenting
No Data leakage
Difficulty: Normal -5
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe acceptable/good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good
